witwidget/notebook/colab/wit.py

```diff
@@ -186,18 +204,34 @@ def compute_custom_distance(wit_id, index, callback_name, params):
           wit.customDistanceFunctionSet = false;
         }}
       }};
+      window.updateExamplesStartCallback = () => {{
+        stagedExamples = [];
+      }};
       window.updateExamplesCallback = examples => {{
+        stagedExamples = stagedExamples.concat(examples);
+      }};
+      window.updateExamplesEndCallback = () => {{
         if (!wit.updateExampleContents) {{
-          requestAnimationFrame(() => window.updateExamplesCallback(examples));
+          requestAnimationFrame(() => window.updateExamplesEndCallback());
           return;
         }}
-        wit.updateExampleContents(examples, false);
+        wit.updateExampleContents(stagedExamples, false);
         if (wit.localAtlasUrl) {{
           window.spriteCallback(wit.localAtlasUrl);
         }}
       }};
-      // BroadcastChannel allows examples to be updated by a call from an
+      // BroadcastChannels allows examples to be updated by a call from an
       // output cell that isn't the cell hosting the WIT widget.
+      const channelStartName = 'updateExamplesStart' + id;
```

Comment:
This is a general design decision but I think it may be better to indicate to UI how many messages are remaining somehow. If you for example sent the number of remaining packets with the packet itself then you would not need updateExamplesEnd callback. The frontend would now when you send a packet with 0 remaining. You probably don't need explicit start channel in that case either as you could always clear staged examples when 0 packet is sent and whenever the new transfer starts it would start fresh.
I think it may be wise to not allow a new transfer start before previous one ends by calling set_examples from a separate cell during transport by having a transfer_lock like variable here.

Reply:
i think in these initial PRs, sending info to frontend that indicates remaining work to be done is a good idea, but we can save putting that on the screen to another change, to keep the changes relatively small and focused

Reply:
also agree with the transfer lock concept of not allowing a set_examples call while one is in progress.